pkg/lib/crd/storage.go

```diff
@@ -0,0 +1,66 @@
+package crd
+
+import (
+	"fmt"
+	apiextensionsv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	apiextensionsv1beta1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1"
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+// SafeStorageVersionUpgrade determines whether the new CRD spec includes all the storage versions of the existing on-cluster CRD.
+// For each stored version in the status of the CRD on the cluster (there will be at least one) - each version must exist in the spec of the new CRD that is being installed.
+// See https://kubernetes.io/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definition-versioning/#upgrade-existing-objects-to-a-new-stored-version.
+func SafeStorageVersionUpgrade(existingCRD runtime.Object, newCRD runtime.Object) error {
+	newSpecVersions, existingStatusVersions := getStoredVersions(existingCRD, newCRD)
+	if newSpecVersions == nil {
+		return fmt.Errorf("could not find a stored version in the spec of the new CRD: found %#v", newSpecVersions)
+	}
+	if existingStatusVersions == nil {
+		return fmt.Errorf("could not find a stored version in the status of the existing CRD: found %#v", existingStatusVersions)
+	}
+
+	for name := range existingStatusVersions {
+		if _, ok := newSpecVersions[name]; !ok {
+			// a storage version in the status of the old CRD is not present in the spec of the new CRD
+			// potential data loss of CRs without a storage migration - throw error and block the CRD upgrade
+			return fmt.Errorf("new CRD removes version %s that is listed as a stored version on the existing CRD", name)
+		}
+	}
+
+	return nil
+}
+
+// getStoredVersions returns the storage versions listed in the status of the old on-cluster CRD
+// and the storage versions listed in the spec of the new CRD.
```

Comment:
Is this correct? The code return all versions in the spec of new CRD, not just storage version.

Reply:
I changed this recently (when writing the unit tests) to pull all versions -- I think this is correct because there can only be one version set to the storage version on the spec of the CRD. So if we only look at storage version we only compare one version to the one or more versions on the status of the existing CRD.

I think we want to confirm all versions stored in the CRD on-cluster are present in the new CRD (many to many relationship).

Reply:
Then please update the comment to reflect that all versions, not just storage versions.